Feature: use token informations (#1585)
stellio-hub/stellio-context-broker: docs/admin/authentication_integration.md

# Authentication integration

> ## Choosing an OIDC provider

Reviewer: ```suggestion
## Choose an OIDC provider
```

> Stellio can be configured to use any OIDC provider and works with all ngsi-ld and permissions endpoints.

Reviewer: "Stellio ... works with all NGSI-LD endpoints" ? 🤔

> If you use the [EGM Keycloak image](#integrate-authentication-with-egm-keycloak-image).

Reviewer: Weird sentence.

The EGM image gives you access to the [subject endpoints](../user/authentication_and_authorization.md#subjects-endpoints)

>  and allow the [`on-owner-delete-cascade-entities` property](../user/authentication_and_authorization.md#delete-entities-owned-by-stellio-user-if-said-user-is-deleted).

Reviewer: ```suggestion
 and allows the [`on-owner-delete-cascade-entities` property](../user/authentication_and_authorization.md#delete-entities-owned-by-stellio-user-if-said-user-is-deleted).
```

Reviewer: better to allow a feature than a property

## Common configuration

> ## Configure the search-service

Reviewer: Heading level problem?

The following environment variables are needed to configure authentication in the search service:
- `STELLIO_AUTHENTICATION_ENABLED`: true (a boolean)
- `APPLICATION_TENANTS_0_ISSUER`: `your-issuer-url/uri`
  -  The URL of the OIDC issuer.

>   -  For example, a keycloak issuer will be  `http://{keycloak_url}/realms/{realm_name}`

Reviewer: ```suggestion
  -  For example, a Keycloak issuer will be  `https://{keycloak_url}/realms/{realm_name}`
```

- `STELLIO_AUTHENTICATION_CLAIMS_PATHS`: `my.claim.path,my.other.claim.path` 
  - A comma-separated list of paths, the path should correspond to a claim in the JWT generated by your OIDC provider.
  - When [evaluating permissions](../user/authentication_and_authorization.md/#manage-access-rights),
    the specified claims will be used to see if the user gets access to the resource.
  - default value is : `realm_access.roles,groups_uuids`

## Configure the subscription-service
The subscription service uses the same base variables as the search service.

Additionally, the subscription service uses client-credentials to authenticate on the search service.
You should create a service account with the "stellio-admin" claim in your OIDC provider.

> With this service account, the following variables can be defined:

Reviewer: ```suggestion
With this service account, the following additional variables must be defined:
```

- `APPLICATION_TENANTS_0_CLIENTID`: `{my-client_id}`
- `APPLICATION_TENANTS_0_CLIENTSECRET`: `{my-client-secret}`
- `APPLICATION_TENANTS_0_ACCESSTOKENURL`: `{my-access-token-url}`

>   - The url of the oidc provider access token url.

Reviewer: ```suggestion
  - The access token URL of the OIDC provider.
```

>   - By default, it uses the keycloak token url `{issuer}/protocol/openid-connect/token`

Reviewer: ```suggestion
  - By default, it uses the Keycloak token URL `{issuer}/protocol/openid-connect/token`
```

## Validate the configuration

To validate the configuration is fully working, the client credentials can be used to create an entity in Stellio:

- Get an access token for this client

```shell
export ACCESS_TOKEN=$(http --form POST {my-access-token-url} client_id={client_id} client_secret={client_secret} grant_type=client_credentials | jq -r .access_token)
```

- Create a sample entity

```shell
echo -n '{ "id": "urn:ngsi-ld:Entity:01", "type": "Entity" }' | http POST http://localhost:8080/ngsi-ld/v1/entities Authorization:"Bearer $ACCESS_TOKEN"
```

If everything is correctly configured, Stellio returns a `201 Created` response.

## Integrate authentication with EGM Keycloak image

> Keycloak integration

Reviewer: to be removed?
